password/src/Data/Password/Argon2.hs

```diff
@@ -0,0 +1,338 @@
+import Data.ByteString (ByteString)
+import Data.ByteString.Base64 (encodeBase64)
+import qualified Data.ByteString.Char8 as C8 (length)
+import Data.Maybe (fromMaybe)
+#if MIN_VERSION_base(4,9,0)
+import Data.Semigroup ((<>))
+#endif
+import Data.Text (Text)
+import qualified Data.Text as T (intercalate, length, split, splitAt)
+import Data.Word (Word32)
+
+import Data.Password (
+         PasswordCheck(..)
+       , PasswordHash(..)
+       , Salt(..)
+       , mkPassword
+       , unsafeShowPassword
+       , unsafeShowPasswordText
+       )
+import Data.Password.Internal (Password(..), from64, readT, showT, toBytes)
+import qualified Data.Password.Internal (newSalt)
+
+
+-- | Phantom type for __Argon2__
+--
+-- @since 2.0.0.0
+data Argon2
+
+-- $setup
+-- >>> :set -XFlexibleInstances
+-- >>> :set -XOverloadedStrings
+--
+-- Import needed libraries.
+--
+-- >>> import Data.Password
+-- >>> import Data.ByteString (pack)
+-- >>> import Test.QuickCheck (Arbitrary(arbitrary), Blind(Blind), vector)
+-- >>> import Test.QuickCheck.Instances.Text ()
+--
+-- >>> instance Arbitrary (Salt a) where arbitrary = Salt . pack <$> vector 16
+-- >>> instance Arbitrary Password where arbitrary = fmap Password arbitrary
+-- >>> let testParams = defaultParams {argon2TimeCost = 1}
+-- >>> let salt = Salt "abcdefghijklmnop"
+--
+-- -- >>> instance Arbitrary (PasswordHash Argon2) where arbitrary = hashPasswordWithSalt testParams <$> arbitrary <*> arbitrary
+
+-- | Hash the 'Password' using the /Argon2/ hash algorithm
+--
+-- >>> hashPassword $ mkPassword "foobar"
+-- PasswordHash {unPasswordHash = "$argon2id$v=19$m=65536,t=2,p=1$...$..."}
+hashPassword :: MonadIO m => Password -> m (PasswordHash Argon2)
+hashPassword = hashPasswordWithParams defaultParams
+
+-- | Parameters used in the /Argon2/ hashing algorithm.
+--
+-- @since 2.0.0.0
+data Argon2Params = Argon2Params {
+  argon2Salt :: Word32,
+  -- ^ Bytes to randomly generate as a unique salt, default is __16__
+  --
+  -- Limits are min: @8@, and max: @2 ^ 32 -1@
+  argon2Variant :: Argon2.Variant,
+  -- ^ Which variant of Argon2 to use ('Argon2d', 'Argon2i' or 'Argon2id')
+  argon2Version :: Argon2.Version,
+  -- ^ Which version of Argon2 to use ('Version10' or 'Version13')
+  argon2MemoryCost :: Word32,
+  -- ^ Memory cost, default is __65536__ (i.e. 64MB)
+  --
+  -- Limits are min: @8 * 'argon2Parallelism'@, and max is addressing
+  -- space / 2, or @2 ^ 32 - 1@, whichever is lower.
+  argon2TimeCost :: Word32,
+  -- ^ Amount of computation realized, default is __2__
+  --
+  -- Limits are min: @1@, and max: @2 ^ 32 - 1@
+  argon2Parallelism :: Word32,
+  -- ^ Parallelism factor, default is __1__
+  --
+  -- Limits are min: @1@, and max: @2 ^ 24 - 1@
+  argon2OutputLength :: Word32
+  -- ^ Output key length in bytes, default is __32__
+  --
+  -- Limits are min: @4@, and max: @2 ^ 32 - 1@
+} deriving (Eq, Show)
+
+-- | Default parameters for the /Argon2/ algorithm.
+--
```

Comment:
I think it is nice to have this explicitly in the haddocks.

```suggestion
--
-- >>> defaultParams
-- Argon2Params {argon2Salt = 16, argon2Variant = Argon2id, argon2Version = Version13, argon2MemoryCost = 65536, argon2TimeCost = 2, argon2Parallelism = 1, argon2OutputLength = 32}
--
```

Reply:
It'd be nice to have this in the other modules with params as well.

Reply:
That's a good point.

Reply:
Updated